kbase-extension/static/kbase/js/widgets/appWidgets2/validation.js

```diff
@@ -411,22 +411,24 @@ define([
          * (ignoring all other constraints)
          *
          * @param {string} value
-         * @param {object} options
+         * @param {object} constraints
          * - required - boolean
          * - min_length - int
          * - max_length - int
          * - regexp - regex (bare string of form /regex/)
          * - type - string, only available now is "WorkspaceObjectName" - if that's present, this
          *      gets validated as a workspace object name (meaning that string lengths and regexp
          *      are ignored)
+         * @param {object} options (optional). if present, optional values are
+         * - invalidValues {Set<string>} set of values that are always invalid
+         * - invalidError {string} string to respond with if an invalid value is present
          */
-        function validateText(value, constraints) {
+        function validateTextString(value, constraints, options = {}) {
```

Comment:
Adjustments here, since, of course, the `validations/text` validator was ever so slightly different. This merges the two, adding the `invalidValues` bit for dropdowns, that we added as part of the bulk import cell. Dropdowns (both static and dynamic) use this validator, and optionally have a Set of values that are invalid choices.